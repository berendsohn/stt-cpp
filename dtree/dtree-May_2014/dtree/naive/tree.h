// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_TREE_H_
#define DTREE_NAIVE_TREE_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndTree : public Base {
 public:
  EndTree() : parent_(NULL) {}
  EndTree* parent() const { return parent_; }
  void set_parent(EndTree* parent) { parent_ = parent; }
  template<typename Forwarder>
  void CopyFrom(const EndTree& node, Forwarder forward);
 private:
  EndTree* parent_;
  DISALLOW_COPY_AND_ASSIGN(EndTree);
};

template<typename Type, typename Base>
class WithStaticValue : public Base {
 public:
  typedef dtree::Nop<Type> Group_;
  Type value() const { return value_; }
  void set_value(Type value) { value_ = value; }
  //
  template<typename Node>
  static Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, Type value);
 protected:
  WithStaticValue() : value_() {}
 private:
  Type value_;
};

template<typename Group, typename Base>
class WithAncValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  //
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void AddToProperAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromProperAnc(Node* node, typename Group::Type delta);

 protected:
  WithAncValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAncAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  //
  template<typename Node>
  static typename Semigroup::Type AggrAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostAnc(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostAnc(Node* node, const Predicate& predicate);
  template<typename Node>
  static typename Semigroup::Type AggrProperAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostProperAnc(int dir,
                                    Node* node,
                                    const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostProperAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostProperAnc(Node* node, const Predicate& predicate);

 protected:
  WithAncAggr() {}
};

template<typename Base>
class WithEvertBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvertBy() {}
};

template<typename Base>
class WithEvert : public Base {
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvert() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_TREE_H_
