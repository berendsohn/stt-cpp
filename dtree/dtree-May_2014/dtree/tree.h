// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_TREE_H_
#define DTREE_TREE_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/tree.h"
#else
#include "dtree/selfadjust/tree.h"
#endif
#endif  // DTREE_TREE_H_
