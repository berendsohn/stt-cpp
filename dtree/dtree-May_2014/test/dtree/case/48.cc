#define DTREE_CONFIG EndTreeWithDesc<WithEvert<WithAncValue<FreeGroup, WithDescAggrOfDescValue<dtree::CountAndSum<int, int>, NullPredicate_, WithDescAggrOfDescValue<dtree::Or<int>, dtree::NonzeroAnd_<int>, WithDescValueWithDesc<dtree::Nop<int>, Begin > > > > > >
#include "test/dtree/tree.cc"
