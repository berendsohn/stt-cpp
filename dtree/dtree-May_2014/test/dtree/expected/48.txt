<<<
  1: v1 v6ca,000,107,000
  2: v3 v744,000,3ca,000 -> 8
  3: v1 v2fe,000,79b,000
  4: v0 v1ac,000,20f,000 -> 12
  5: v7 v586,000,e4d,000 -> 14
  6: v0 veb0,000,96d,000
  7: v2 vb53,000,aa5,000 -> 8
  8: v5 vcd5,000,f6c,000 -> 14
  9: v2 v1ac,000,74f,000
  10: v4 v12c,000,9ce,000
  11: v4 v334,000,492,000 -> 8
  12: v3 v93b,000,309,000
  13: v1 v4b3,000,f5a,000
  14: v1 v7b3,000,07c,000 -> 3
  15: v5 v8d2,000,623,000 -> 14
>>>
<<<
  @1: v1 v1 a1 v1 v1 a1 aC1S1 v6ca,000,107,000 f0

  @6: v0 v0 a0 v0 v0 a0 aC1S0 veb0,000,96d,000 f0

  @9: v2 v2 a2 v2 v2 a2 aC1S2 v1ac,000,74f,000 f1

  @10: v4 v4 a4 v4 v4 a4 aC1S4 v12c,000,9ce,000 f1

    /4: v0 v0 a0 v0 v0 a0 aC1S0 v1ac,93b,20f,309 f0
  @12: v3 v3 a3 v3 v3 a3 aC1S3 v93b,000,309,000 f1

  @13: v1 v1 a1 v1 v1 a1 aC1S1 v4b3,000,f5a,000 f1

  @15: v5 v5 a5 v5 v5 a5 aC1S5 v8d2,000,623,000 f1
    \14: v1 v1 a7 v1 v1 a7 aC7S23 v7b3,8d2,07c,623 f0
       .'5: v7 v7 a7 v7 v7 a7 aC1S7 v586,000,e4d,000 f1
        /11: v4 v4 a4 v4 v4 a4 aC1S4 v334,cd5,492,f6c f1
     ->8: v5 v5 a7 v5 v5 a7 aC5S21 vcd5,000,f6c,000 f1
         .'2: v3 v3 a3 v3 v3 a3 aC1S3 v744,000,3ca,000 f1
       ->7: v2 v2 a3 v2 v2 a3 aC2S5 vb53,000,aa5,000 f1
      \3: v1 v1 a1 v1 v1 a1 aC1S1 v2fe,7b3,79b,07c f0
>>>
